prow/git/git_command_test.go

```diff
@@ -0,0 +1,273 @@
+			},
+			expected: Repo{
+				Dir:  tempdir.recorded(1),
+				git:  gitbin,
+				base: urlMustParse("https://github.com"),
+				repo: "org/repo",
+				user: "user",
+				pass: "password",
+			},
+		},
+		{
+			name:                 "happy path with remote url and user, one retry",
+			base:                 "https://github.com",
+			user:                 "user",
+			tokenGenerator:       func() []byte { return []byte("password") },
+			repo:                 "org/repo",
+			simulateExecFailures: 1,
+			commands: [][]string{
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(2) + "/org/repo.git"},
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(2) + "/org/repo.git"},
+				{"", gitbin, "clone", tempdir.recorded(2) + "/org/repo.git", tempdir.recorded(3)},
+			},
+			expected: Repo{
+				Dir:  tempdir.recorded(3),
+				git:  gitbin,
+				base: urlMustParse("https://github.com"),
+				repo: "org/repo",
+				user: "user",
+				pass: "password",
+			},
+		},
+		{
+			name:                 "happy path with remote url and user, 3 retries",
+			base:                 "https://github.com",
+			user:                 "user",
+			tokenGenerator:       func() []byte { return []byte("password") },
+			repo:                 "org/repo",
+			simulateExecFailures: 3,
+			commands: [][]string{
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(4) + "/org/repo.git"},
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(4) + "/org/repo.git"},
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(4) + "/org/repo.git"},
+				{"", gitbin, "clone", tempdir.recorded(4) + "/org/repo.git", tempdir.recorded(5)},
+			},
+			expected: Repo{
+				Dir:  tempdir.recorded(5),
+				git:  gitbin,
+				base: urlMustParse("https://github.com"),
+				repo: "org/repo",
+				user: "user",
+				pass: "password",
+			},
+		},
+		{
+			name:                 "error path with remote url and user, retries exceeded",
+			base:                 "https://github.com",
+			user:                 "user",
+			tokenGenerator:       func() []byte { return []byte("password") },
+			repo:                 "org/repo",
+			simulateExecFailures: 4,
+			commands: [][]string{
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(6) + "/org/repo.git"},
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(6) + "/org/repo.git"},
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(6) + "/org/repo.git"},
+				{"", gitbin, "clone", "--mirror", "https://user:password@github.com/org/repo", tempdir.recorded(6) + "/org/repo.git"},
+			},
+			err: true,
+		},
+		{
+			name:                 "happy path with GHE",
```

Comment:
It's really hard for me to understand the semantic difference between test cases here, there is a lot going on and it's not obvious which parts are different between each

Reply:
This also means the tests are likely fragile and will all fail if someone makes Clone do something new